PEP 674: Leave PyDescr_TYPE() unchanged (#2277)
python/peps: pep-0674.rst

* Ease the C API implementation on other Python implementation;
* Help migrating existing C extensions to the HPy API.

> On the PyPI top 5000 projects, only 13 projects (0.3%) are affected by 2

Reviewer: ```suggestion
Only 13 out of the top 5000 projects on PyPI (0.3%) are affected by 2
```

Reviewer 2: There is a minor grammar and repetition issue here, but @vstinner here's a more conservative change if you prefer something closer to the original wording (@AA-Turner 's suggestion is fine too):

```suggestion
Of the top 5000 PyPI projects, only 13 (0.3%) are affected by 2
```

It might be helpful to clarify the "top 5000 _most downloaded_" given there can be different measures of "top"-ness, but perhaps I doth protest too much

```suggestion
Of the top 5000 most-downloaded PyPI projects, only 13 (0.3%) are affected by 2
```

macro changes. Moreover, 22 projects just have to regenerate their

> Cython code.

Reviewer 2: As a reader of the PEP, it confused me as to how there could be 22 projects that had to re-Cythonize, but only 13 that were affected. What I _think_ you mean is 

```suggestion
macro changes. An additional 22 projects just have to regenerate their
Cython code.
```

Reviewer 2: Superceded by followup comment

file which provides ``Py_SET_REFCNT()``, ``Py_SET_TYPE()`` and
``Py_SET_SIZE()`` functions to Python 3.8 and older.

Backwards Compatibility
=======================

The proposed C API changes are backward incompatible on purpose.

On January 26, 2022, a code search on the top 5000 PyPI projects (4762
projects in practice, others don't have a source archive) found that
only 13 projects are affected (0.3%):

* Cython (0.29.26)

> * PyGObject (3.42.0)

Reviewer 2: Mooted by #2280

* datatable (1.0.0)
* guppy3 (3.1.2)
* numpy (1.22.1)
* pickle5 (0.0.12)
* psycopg2 (2.9.3)

> * pycurl (7.44.1)

Reviewer 2: Mooted by #2280

file which provides ``Py_SET_REFCNT()``, ``Py_SET_TYPE()`` and
``Py_SET_SIZE()`` functions to Python 3.8 and older.

Backwards Compatibility
=======================

The proposed C API changes are backward incompatible on purpose.

> At January 26, 2022, a code search on the PyPI top 5000 projects (4762

Reviewer 3: ```suggestion
On January 26, 2022, a code search on the top 5000 PyPI projects (4762
```

> projects in practice, others don't have a source archive) found that

Reviewer 2: ```suggestion
projects in practice; others don't have a source archive) found that
```

Comma splice (grammar error)

> only 13 projects are affected (0.3%):

Reviewer: This duplicates the above -- I'd merge them. I'm not sure the line around the 238 projects without sdists is needed.

* Cython (0.29.26)
* PyGObject (3.42.0)
* datatable (1.0.0)
* guppy3 (3.1.2)
* numpy (1.22.1)
* pickle5 (0.0.12)
* psycopg2 (2.9.3)
* pycurl (7.44.1)
* pysha3 (1.0.2)
* python-snappy (0.6.0)
* recordclass (0.17.1)
* scipy (1.7.3)
* zodbpickle (2.2.0)

> On these 13 projects, only 2 macros are used as l-value: ``Py_TYPE()``

Reviewer: ```suggestion
On these 13 projects, only 2 macros are used as an l-value: ``Py_TYPE()``
```

Reviewer 3: And "Of...":
```suggestion
Of these 13 projects, only 2 macros are used as an l-value: ``Py_TYPE()``
```

and ``Py_SIZE()``.

> Moreover, 22 projects just have to regenerate their Cython code to use

Reviewer 2: ```suggestion
An additional 22 projects just have to regenerate their Cython code to use
```

As above, I'd suggest clarifying this as it confused me on my first reading of the PEP

``Py_SET_TYPE()`` and ``Py_SET_SIZE()``.

This change does not follow the :pep:`387` deprecation process. There is no
known way to emit a deprecation warning only when a macro is used as a
l-value, but not when it's used differently (ex: as a r-value).

PyTuple_GET_ITEM() and PyList_GET_ITEM() are left unchanged
-----------------------------------------------------------

The ``PyTuple_GET_ITEM()`` and ``PyList_GET_ITEM()`` macros are left
unchanged.

> The code pattern ``&PyTuple_GET_ITEM(tuple, 0)`` and
> ``&PyList_GET_ITEM(list, 0)`` is still commonly used to get access to

Reviewer 3: ```suggestion
The code patterns ``&PyTuple_GET_ITEM(tuple, 0)`` and
``&PyList_GET_ITEM(list, 0)`` are still commonly used to get access to
```

the inner ``PyObject**`` array.

Changing these macros is out of the scope of this PEP.

PyDescr_NAME() and PyDescr_TYPE() are left unchanged
----------------------------------------------------

The ``PyDescr_NAME()`` and ``PyDescr_TYPE()`` macros are left unchanged.

> The ``PyDescr_NAME()`` and ``PyDescr_TYPE()`` macros give access to

Reviewer 2: ```suggestion
These macros
```

Avoid word for word repetition of the previous sentence.

``PyDescrObject.d_name`` and ``PyDescrObject.d_type`` members. They can

> be used as l-value to set these members.

Reviewer 3: ```suggestion
be used as l-values to set these members.
```

> The SWIG project use these macros as l-value to set these members. It

Reviewer 3: ```suggestion
The SWIG project uses these macros as l-values to set these members. It
```

> would be possible to modify SWIG to prevent setting directly
> ``PyDescrObject`` structure members, but it is not really worth it since

Reviewer: ```suggestion
would be possible to modify SWIG to prevent setting ``PyDescrObject``
structure members directly, but it is not really worth it since
```

> the ``PyDescrObject`` structure is not performance criticial and is

Reviewer 4: ```suggestion
the ``PyDescrObject`` structure is not performance critical and is
```

the ``Py_TYPE()`` documentation now requires to use the
``Py_SET_TYPE()`` function to set an object type.

If developers use macros as an l-value, it's their responsibility when

> their code breaks, not the Python responsibility. We are operating under

Reviewer 2: ```suggestion
their code breaks, not Python's responsibility. We are operating under
```

Grammar

so are not aware of compatibility issues with other Python
implementations.

Moreover, continuing to allow using macros as an l-value does not help

> the HPy project and leaves the burden of emulating them on GraalVM's

Reviewer 2: ```suggestion
the HPy project, and leaves the burden of emulating them on GraalVM's
```

Comma needed here, both syntactically and semantically

  <https://bugs.python.org/issue30459>`_ (May 2017)

Version History
===============

* Version 3: No longer change PyDescr_TYPE() and PyDescr_NAME() macros

> * Version 2: Add "Relationship with the HPy proejct" section, remove

Reviewer 2: ```suggestion
* Version 2: Add "Relationship with the HPy project" section; remove
```

Fix typo and comma splice
